### PR TITLE
Fix 'undefined' bucket type issue.

### PR DESCRIPTION
**Pull request by seancribbs**
This was broken by #970. The client is not required to send a bucket-type in the request, therefore it can be the atom 'undefined', which corresponds to the "default" bucket-type. Also untangled the logic into a separate "check" for the presence of the bucket-type.

This was found by the Python client suite here: http://buildbot.bos1/builders/python-client-test/builds/60/steps/make%20test/logs/stdio

Also updated tools.mk.

@kellymclaughlin

### CODE REVIEW

**Comment by seancribbs** [comment]
Just in case you don't want to mess with the VPN, I've pasted the failing output from the test suite below. FWIW, 'pytest' is a real bucket-type that was created and activated before the suite ran, and should not result in a failure.

``````
======================================================================
ERROR: test_btype_list_buckets (riak.tests.test_all.RiakPbcTransportTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/buildbot/masters/riak/python-client-test/build/riak/tests/test_btypes.py", line 88, in test_btype_list_buckets
    self.assertIn(bucket, btype.get_buckets())
  File "/home/buildbot/masters/riak/python-client-test/build/riak/bucket.py", line 627, in get_buckets
    return self._client.get_buckets(bucket_type=self, timeout=timeout)
  File "/home/buildbot/masters/riak/python-client-test/build/riak/client/transport.py", line 184, in wrapper
    return self._with_retries(pool, thunk)
  File "/home/buildbot/masters/riak/python-client-test/build/riak/client/transport.py", line 126, in _with_retries
    return fn(transport)
  File "/home/buildbot/masters/riak/python-client-test/build/riak/client/transport.py", line 182, in thunk
    return fn(self, transport, *args, **kwargs)
  File "/home/buildbot/masters/riak/python-client-test/build/riak/client/operations.py", line 62, in get_buckets
    timeout=timeout)]
  File "/home/buildbot/masters/riak/python-client-test/build/riak/transports/pbc/transport.py", line 292, in get_buckets
    MSG_CODE_LIST_BUCKETS_RESP)
  File "/home/buildbot/masters/riak/python-client-test/build/riak/transports/pbc/connection.py", line 50, in _request
    return self._recv_msg(expect)
  File "/home/buildbot/masters/riak/python-client-test/build/riak/transports/pbc/connection.py", line 143, in _recv_msg
    raise RiakError(err.errmsg)
RiakError: "No bucket-type named 'pytest'"

======================================================================
ERROR: test_default_btype_list_buckets (riak.tests.test_all.RiakPbcTransportTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/buildbot/masters/riak/python-client-test/build/riak/tests/test_btypes.py", line 119, in test_default_btype_list_buckets
    self.assertIn(bucket, default_btype.get_buckets())
  File "/home/buildbot/masters/riak/python-client-test/build/riak/bucket.py", line 627, in get_buckets
    return self._client.get_buckets(bucket_type=self, timeout=timeout)
  File "/home/buildbot/masters/riak/python-client-test/build/riak/client/transport.py", line 184, in wrapper
    return self._with_retries(pool, thunk)
  File "/home/buildbot/masters/riak/python-client-test/build/riak/client/transport.py", line 126, in _with_retries
    return fn(transport)
  File "/home/buildbot/masters/riak/python-client-test/build/riak/client/transport.py", line 182, in thunk
    return fn(self, transport, *args, **kwargs)
  File "/home/buildbot/masters/riak/python-client-test/build/riak/client/operations.py", line 62, in get_buckets
    timeout=timeout)]
  File "/home/buildbot/masters/riak/python-client-test/build/riak/transports/pbc/transport.py", line 292, in get_buckets
    MSG_CODE_LIST_BUCKETS_RESP)
  File "/home/buildbot/masters/riak/python-client-test/build/riak/transports/pbc/connection.py", line 50, in _request
    return self._recv_msg(expect)
  File "/home/buildbot/masters/riak/python-client-test/build/riak/transports/pbc/connection.py", line 143, in _recv_msg
    raise RiakError(err.errmsg)
RiakError: "No bucket-type named 'undefined'"

======================================================================
ERROR: test_list_buckets (riak.tests.test_all.RiakPbcTransportTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/buildbot/masters/riak/python-client-test/build/riak/tests/test_kv.py", line 454, in test_list_buckets
    buckets = self.client.get_buckets()
  File "/home/buildbot/masters/riak/python-client-test/build/riak/client/transport.py", line 184, in wrapper
    return self._with_retries(pool, thunk)
  File "/home/buildbot/masters/riak/python-client-test/build/riak/client/transport.py", line 126, in _with_retries
    return fn(transport)
  File "/home/buildbot/masters/riak/python-client-test/build/riak/client/transport.py", line 182, in thunk
    return fn(self, transport, *args, **kwargs)
  File "/home/buildbot/masters/riak/python-client-test/build/riak/client/operations.py", line 62, in get_buckets
    timeout=timeout)]
  File "/home/buildbot/masters/riak/python-client-test/build/riak/transports/pbc/transport.py", line 292, in get_buckets
    MSG_CODE_LIST_BUCKETS_RESP)
  File "/home/buildbot/masters/riak/python-client-test/build/riak/transports/pbc/connection.py", line 50, in _request
    return self._recv_msg(expect)
  File "/home/buildbot/masters/riak/python-client-test/build/riak/transports/pbc/connection.py", line 143, in _recv_msg
    raise RiakError(err.errmsg)
RiakError: "No bucket-type named 'undefined'"

======================================================================
ERROR: test_stream_buckets (riak.tests.test_all.RiakPbcTransportTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/buildbot/masters/riak/python-client-test/build/riak/tests/test_kv.py", line 462, in test_stream_buckets
    for bucket_list in self.client.stream_buckets():
  File "/home/buildbot/masters/riak/python-client-test/build/riak/client/operations.py", line 89, in stream_buckets
    for bucket_list in stream:
  File "/home/buildbot/masters/riak/python-client-test/build/riak/transports/pbc/stream.py", line 118, in next
    response = super(RiakPbcBucketStream, self).next()
  File "/home/buildbot/masters/riak/python-client-test/build/riak/transports/pbc/stream.py", line 51, in next
    msg_code, resp = self.transport._recv_msg(expect=self._expect)
  File "/home/buildbot/masters/riak/python-client-test/build/riak/transports/pbc/connection.py", line 143, in _recv_msg
    raise RiakError(err.errmsg)
RiakError: "No bucket-type named 'undefined'"

======================================================================
ERROR: test_stream_buckets_abort (riak.tests.test_all.RiakPbcTransportTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/buildbot/masters/riak/python-client-test/build/riak/tests/test_kv.py", line 472, in test_stream_buckets_abort
    for bucket_list in self.client.stream_buckets():
  File "/home/buildbot/masters/riak/python-client-test/build/riak/client/operations.py", line 89, in stream_buckets
    for bucket_list in stream:
  File "/home/buildbot/masters/riak/python-client-test/build/riak/transports/pbc/stream.py", line 118, in next
    response = super(RiakPbcBucketStream, self).next()
  File "/home/buildbot/masters/riak/python-client-test/build/riak/transports/pbc/stream.py", line 51, in next
    msg_code, resp = self.transport._recv_msg(expect=self._expect)
  File "/home/buildbot/masters/riak/python-client-test/build/riak/transports/pbc/connection.py", line 143, in _recv_msg
    raise RiakError(err.errmsg)
RiakError: "No bucket-type named 'undefined'"

----------------------------------------------------------------------```
``````

**Comment by kellymclaughlin** [comment]
I'm still getting these two python test failures, but they seem to be unrelated to this change:

```
======================================================================
FAIL: test_clear_bucket_properties (riak.tests.test_all.RiakHttpTransportTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/kelly/basho/repos/riak-python-client/riak/tests/test_kv.py", line 549, in test_clear_bucket_properties
    self.assertFalse(bucket.allow_mult)
AssertionError: True is not false

======================================================================
FAIL: test_clear_bucket_properties (riak.tests.test_all.RiakPbcTransportTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/kelly/basho/repos/riak-python-client/riak/tests/test_kv.py", line 549, in test_clear_bucket_properties
    self.assertFalse(bucket.allow_mult)
AssertionError: True is not false

----------------------------------------------------------------------
Ran 301 tests in 43.083s

FAILED (failures=2, skipped=73)
```

File: src/riak_kv_pb_bucket.erl [hunk 1]
```diff
@@ -167,25 +159,29 @@ process_stream({ReqId, Error}, ReqId,
...
--spec maybe_bucket_type_exists(BucketExists :: atom() | binary() | list()) -> atom().
-maybe_bucket_type_exists(undefined) ->
-    false;
-maybe_bucket_type_exists(Type) when is_list(Type) ->
-    true;
-maybe_bucket_type_exists(Type) when is_binary(Type) ->
-    maybe_bucket_type_exists(riak_core_bucket_type:get(Type)).
 
--spec maybe_create_bucket_type(Type :: binary() | atom(), Bucket :: binary()) -> any() | { binary(), list() } | { atom() | atom() }.
+do_list_buckets(Type, Timeout, true, Req, #state{client=C}=State) ->
```
Comment (kellymclaughlin) [inline src/riak_kv_pb_bucket.erl:167]: How about adding specs to these new functions?

**Comment by kellymclaughlin** [comment]
- [x] eunit & eqc
- [x] python client tests (caveat is the 2 unrelated failures noted above)
- [x] dialyzer
- [x] xref
- [x] verify_listkeys riak_test

File: src/riak_kv_pb_bucket.erl (hunk 1 above)
Comment (seancribbs) [inline src/riak_kv_pb_bucket.erl:167]: Addressed in f07afe2

**Comment by seancribbs** [comment]
We should wait until #976 merges so we can get the whole suite.

**Comment by kellymclaughlin** [comment]
Eunit and eqc test pass when rebasing the branch from latest `develop`. Also re-ran dialyzer to verify new specs. Looks good!

:+1:

**Comment by seancribbs** [comment]
@borshop merge

**Pull request merged**